dll/win32/shell32/CCopyToMenu.cpp

```diff
@@ -0,0 +1,173 @@
+#include "precomp.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+CCopyToMenu::CCopyToMenu() :
+    m_idCmdFirst(0),
+    m_idCmdLast(0),
+    m_idCmdCopyTo(-1)
+{
+}
+
+CCopyToMenu::~CCopyToMenu()
+{
+}
+
+static int CALLBACK
+BrowseCallbackProc(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
+{
+    CCopyToMenu *this_ =
+        reinterpret_cast<CCopyToMenu *>(GetWindowLongPtr(hwnd, GWLP_USERDATA));
+    LPCITEMIDLIST pidl;
+    WCHAR szPath[MAX_PATH];
+
+    switch (uMsg)
+    {
+        case BFFM_INITIALIZED:
+            SetWindowLongPtr(hwnd, GWLP_USERDATA, lpData);
+            this_ = reinterpret_cast<CCopyToMenu *>(lpData);
+
+            // Select initial directory
+            SendMessageW(hwnd, BFFM_SETSELECTION, FALSE,
+                reinterpret_cast<LPARAM>(static_cast<LPCITEMIDLIST>(this_->m_pidlFolder)));
+            break;
+
+        case BFFM_SELCHANGED:
+            pidl = reinterpret_cast<LPCITEMIDLIST>(lParam);
+            if (SHGetPathFromIDListW(pidl, szPath) && PathFileExistsW(szPath) &&
+                !ILIsEqual(pidl, this_->m_pidlFolder))
+            {
+                SendMessageW(hwnd, BFFM_ENABLEOK, 0, TRUE);
+            }
+            else
+            {
+                SendMessageW(hwnd, BFFM_ENABLEOK, 0, FALSE);
+            }
+            break;
+    }
+
+    return FALSE;
+}
+
+HRESULT CCopyToMenu::DoCopyToFolder(LPCMINVOKECOMMANDINFO lpici)
+{
+    WCHAR wszPath[MAX_PATH];
+    HRESULT hr;
+
+    hr = SHGetPathFromIDListW(m_pidlFolder, wszPath);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    CStringW strTitle(MAKEINTRESOURCEW(IDS_COPYTOTITLE));
+
+    BROWSEINFOW info = { lpici->hwnd };
+    info.pidlRoot = NULL;
+    info.lpszTitle = strTitle;
+    info.ulFlags = BIF_RETURNONLYFSDIRS | BIF_USENEWUI;
+    info.lpfn = BrowseCallbackProc;
+    info.lParam = (LPARAM)this;
+    CComHeapPtr<ITEMIDLIST> pidl(SHBrowseForFolder(&info));
+    if (pidl)
+    {
+        // FIXME: Copy m_pDataObject into pidl
```

Comment:
@learn-more How to copy `m_pDataObject`?

Reply:
Cannot copy from data object to pidl 1:1
A data object can contain multiple files, afaik a single pidl not.

Reply:
Or do you want to create a data object from a pidl?

Reply:
Then how to drop the data object onto the pidl?

Reply:
You probably want something like `CFontExt::Drop` is doing:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/shellext/fontext/CFontExt.cpp;hb=03e6eca3c2aaa61059085570f8e14039f401158f#l475

Reply:
or if you mean apply the copy action for all files in the data object to pidl, then maybe `SHFileOperation`,
where you construct the `pFrom` with the data object items, and the `pTo` from the pidl?